RecoVertex/BeamSpotProducer/plugins/OnlineBeamSpotESProducer.cc

```diff
@@ -51,28 +60,77 @@ OnlineBeamSpotESProducer::OnlineBeamSpotESProducer(const edm::ParameterSet& p) {
 
 void OnlineBeamSpotESProducer::fillDescriptions(edm::ConfigurationDescriptions& desc) {
   edm::ParameterSetDescription dsc;
+  dsc.add<int>("timeThreshold", 48);       // hours
+  dsc.add<double>("sigmaZThreshold", 2.);  // cm
   desc.addWithDefaultLabel(dsc);
 }
 
 const BeamSpotOnlineObjects* OnlineBeamSpotESProducer::compareBS(const BeamSpotOnlineObjects* bs1,
                                                                  const BeamSpotOnlineObjects* bs2) {
-  //Random logic so far ...
-  if (bs1->GetSigmaZ() - 0.0001 > bs2->GetSigmaZ()) {  //just temporary for debugging
-    if (bs1->GetSigmaZ() > 2.5) {
+  // Current time to be compared with the one read from the payload
+  auto currentTime =
+      std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch());
+
+  // Get two beamspot creation times and compute the time difference wrt currentTime
+  auto bs1time = std::chrono::microseconds(bs1->GetCreationTime());
+  auto diffBStime1 = (currentTime - bs1time).count();
+  auto bs2time = std::chrono::microseconds(bs2->GetCreationTime());
+  auto diffBStime2 = (currentTime - bs2time).count();
+
+  // Convert timeThreshold_ from hours to microseconds for comparison
+  auto limitTime = std::chrono::microseconds((std::chrono::hours)timeThreshold_).count();
+
+  // Logic to choose between the two BeamSpots:
+  // 1. If both BS are older than limitTime retun fake BS
+  // 2. If only one BS is newer than limitTime return it only if it has
+  //    sigmaZ larger than sigmaZthreshold_ and the fit converged (BeamType 2)
+  // 3. If both are newer than the limit threshold return
+  //    the BS that converged and has larger sigmaZ
```

Comment:
just for my curiosity, why is the larger sigmaZ better? Shouldn't the logic select a better fit (e.g. smaller `GetSigmaZError`)?

Reply:
if the fit converges to a too small sigmaZ, beamspot compatibility cuts implemented in the downstream POG code (e.g. e/gamma) will start to fail leading to data loss at HLT (at least it was the case up to some point in Run2). I think it was mitigated in the meanwhile.

Reply:
The EGamma code has been fixed, but we cannot be 100% sure the problem would not happen for special runs or different paths.

Reply:
> Is the sigmaZ expected to usually increase or decrease during the fill? I seem to recall that it varies even more due to the lumi leveling.

In Run2 it went up:

![image](https://user-images.githubusercontent.com/5082376/134199951-338dcae7-46b7-4d19-a106-f1af0c02a82d.png)

what does this have to do with the topic under discussion?

Reply:
well I guess Slava was worried we may have cut out good BS values at the end of the fill. But I do not think it is the case.

Reply:
I don't see how that's the case, when you are comparing two different fits of the same LS and not using a previous one (only in the case both fits fail)

Reply:
> I don't see how that's the case, when you are comparing two different fits of the same LS and not using a previous one (only in the case both fits fail)

I may have misunderstood the code then, the comment points to a selection of the best value out of two derived within 48 hours.
If this is all about multiple fits in the same LS, then it doesn't matter indeed.

Reply:
> if you are referring to point 2:

no, I was referring to point 3



> indeed it should be comparing two fits of the same LS.

Thank you for clarifying.